improve(across-v2-umip): Add more details and definitions, and additional rules implemented in Dataworker (#550)
UMAprotocol/UMIPs: UMIPs/umip-157.md

The DVM should support the ACROSS-V2 price identifier.

# Across V2 Architecture

The basic architecture of Across V2 is a single LP ("Liquidity Provider") pool sitting on Ethereum mainnet connected to many "spoke pools" deployed on

> various chains to facilitate user "deposits". A deposit is a cross-chain transfer request from an "origin" chain to a different "destination" chain, which is fulfilled when a "relayer" sends the depositor their desired transfer amount (less fees) on their desired destination chain.

Reviewer: +1

Relayers lend capital to the Across V2 system by fulfilling users' deposits via the spokes and are eventually repaid by the LP pool. "Bundles" containing many of these repayments are validated together by the [Optimistic Oracle ("OO")](https://docs.umaproject.org/protocol-overview/how-does-umas-oracle-work). In addition to validating individual repayment instructions, the OO also validates rebalance instructions that tell the LP pool how to transfer funds to and from the spoke pools in order to execute the repayments and pull deposited funds from the spoke to the LP pool.

If there is no relayer who can provide all the capital for a given deposit request, a "slow relay" (or "slow fill") is performed where the funds are sent from the

> LP pool to the destination spoke to fulfill the deposit. These slow fill requests are also included in the aformentioned Bundles.

Reviewer: ```suggestion
LP pool to the destination spoke to fulfill the deposit. These slow fill requests are also included in the aforementioned bundles.
```

> Bundles are implemented on-chain as [Merkle Roots](https://www.youtube.com/watch?v=JXn4GqyS7Gg) which uniquely identify the set of all repayments and rebalance instructions over a specific time range. Therefore, Across V2 moves capital to repay relayers by continuous Bundles, all validated by the OO.

Reviewer: ```suggestion
Bundles are implemented on-chain as [Merkle Roots](https://www.youtube.com/watch?v=JXn4GqyS7Gg) which uniquely identify the set of all repayments and rebalance instructions over a specific block range. Therefore, Across V2 moves capital to repay relayers and fulfil bridge requests through periodic bundles, all validated by the OO.
```

This UMIP explains exactly how to construct and verify a bundle.

![](./images/across-architecture.png?raw=true "Across V2 Architecture")

# Motivation

The ACROSS-V2 price identifier is intended to be used by the Across v2 contracts to verify whether a bundle of bridge-related
transactions submitted to mainnet is valid.

# Data Specifications and Implementation

Note 1: the following details will often refer to the [Across V2](https://github.com/across-protocol/contracts-v2) repo
at commit hash: a8ab11fef3d15604c46bba6439291432db17e745. This allows the UMIP to have a constant reference rather than
depending on a changing repository.

Note 2: when referencing "later" or "earlier" events, the primary sort should be on the block number, the secondary
sort should be on the `transactionIndex`, and the tertiary sort should be on the `logIndex`. See the section on [comparing events](#comparing-events-chronologically) for more details.

Note 3: wherever unspecified, sorting should be ascending by default, not descending.

# Definitions

## Comparing events chronologically

> Smart contract transactions can emit events that conform to the specifications described in the "Returns" section of these [docs](https://web3js.readthedocs.io/en/v1.2.11/web3-eth-contract.html#contract-events). Specifically, an event is expected to have a `blockNumber`, `transactionIndex` and `logIndex`. To compare events `e1` and `e2` chronologically, we can say

Reviewer: ```suggestion
Smart contract transactions can emit events that conform to the specifications described in the "Returns" section of these [docs](https://web3js.readthedocs.io/en/v1.2.11/web3-eth-contract.html#contract-events). Specifically, an event is expected to have a unique combination of `blockNumber`, `transactionIndex` and `logIndex`. To compare events `e1` and `e2` chronologically, we can say
```

that `e1` is "earlier" than `e2` if `e1.blockNumber < e2.blockNumber` OR if `e1.blockNumber == e2.blockNumber && e1.transactionIndex < e2.transactionIndex` OR if `e1.blockNumber == e2.blockNumber && e1.transactionIndex == e2.transactionIndex && e1.logIndex < e2.logIndex`.

So, "earlier" events have a lower block number, transaction index, or log index, and we should compare event properties in that order.

## Valid bundle proposals

> A valid root bundle submitted by calling `HubPool.proposeRootBundle()` will emit a [`ProposedRootBundle`](https://github.com/across-protocol/contracts-v2/blob/a8ab11fef3d15604c46bba6439291432db17e745/contracts/HubPool.sol#L148-L156) event.

Reviewer: ```suggestion
A root bundle can be proposed by calling `HubPool.proposeRootBundle()`, which will will emit a [`ProposedRootBundle`](https://github.com/across-protocol/contracts-v2/blob/a8ab11fef3d15604c46bba6439291432db17e745/contracts/HubPool.sol#L148-L156) event.
```

   to determine which running balance that this event should apply to. For every [validated root bundle](#valid-bundle-proposals) in history, follow the steps in the ["Finding Slow Relays"](#finding-slow-relays) section
   for this `originChainId` and `depositId`, and look for slow relays with a matching
   `originChainId` and `depositId`. There should be exactly one slow relay payment matching the `FilledRelay` in question. This is the slow fill that was included in a previous bundle that added to the `runningBalance` and ultimately led the bundle to send a slow fill payment to the `SpokePool` on the `destinationChainId`. Compute the [slow fill amount](#computing-slow-relay-payment-amounts) sent in the older root bundle. The excess amount remaining in the `SpokePool` after this current `FilledRelay` (with `isSlowRelay = true`) is equal to `slow fill amount` minus `FilledRelay.fillAmount`. In other words, if the `FilledRelay.fillAmount` is less than the `slow fill amount` originally sent over in a prior bundle, then send back the difference. Subtract the result from the running
   balance for the associated `l1Token` and `destinationChainId`.

> 5. _Subtract excesses from unexecuted slow fills_: This section is similar to the one above but deals with cases where the slow fill leaf can never be executed. Find all `FilledRelay` events in the block range that have `totalFilledAmount` equal to `amount` (i.e. fills that completed a deposit) and that have `fillAmount` less than `amount` (i.e. fills that were not the first fill for a deposit). If the first fill for a deposit completed a deposit (`fillAmount == amount` and `totalFilledAmount == amount`), then there can be no excess left in the spoke pool because this would not have triggered a slow fill payment to be sent to the spoke. First we need to see if the first fill for this current fill triggered a slow fill. Looking through all [valid fills](#valid-bundle-proposals) in history, identify all matching fills for the same `originChainId` and `depositId`. Find the earliest such fill. Determine the block range of the root bundle proposal that contained this fill using [this logic](#determining-block-range-for-root-bundle-proposal) for the `ProposeRootBundle` event with a `bundleEndBlock` for the `FilledRelay.destinationChainId` greater than or equal to the `FilledRelay` block number. Find the last fill in this same bundle block range. The slow fill payment for the bundle should have been `FilledRelay.amount - FilledRelay.totalFilledAmount`, same as [this calculation](#computing-slow-relay-payment-amounts). Since we know that the current `FilledRelay` in this upcoming proposal 100% filled the deposit, we know that the slow fill leaf cannot be executed, so the entire slow fill payment must be sent back to mainnet. Subtract this (the preceiding slow fill payment from a valid root bundle that this fill ultimately completed) from the running balance.

Reviewer: nit: do we specify that this logic should only apply for past bundles?

Author: I think this is implied by this line `Looking through all [valid fills](#valid-bundle-proposals) in history`, as `valid` can only refer to past events, right?

Reviewer: Maybe we should specify "in previously validated bundles"

Author: Done
